File: leetcode/java/0733. Flood Fill.java
package leetcode.java;

// Time complexity: O(E+V), where E are the edges and V are the vertices (grids)
// Space complexity: O(E+V)

class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        this.dfs(image, sr, sc, color, image[sr][sc]);
        return image;
    }

    public void dfs(int[][] image, int row, int col, int color, int prevColor) {

        if (row>=0 && row<image.length && col>=0 && col<image[0].length && image[row][col]==prevColor) {
            image[row][col] = color;
            this.dfs(image, row+1, col, color, prevColor);
            this.dfs(image, row-1, col, color, prevColor);
            this.dfs(image, row, col+1, color, prevColor);
            this.dfs(image, row, col-1, color, prevColor);
        }
    }
}

File: leetcode/python/0733. Flood Fill.py
# Time complexity: O(E+V), where E are the edges and V are the vertices (grids)
# Space complexity: O(E+V)

class Solution(object):
    def floodFill(self, image, sr, sc, color):

        visited = set()
        target = image[sr][sc]

        self.dfs(image, visited, color, sr, sc, target)
        
        return image

    def dfs(self, image, visited, color, row, col, target):
        if (row, col) in visited or row < 0 or col < 0 or row>=len(image) or col>=len(image[0]) or image[row][col] != target:
            return
        visited.add((row, col))
        image[row][col] = color

        self.dfs(image, visited, color, row+1, col, target)
        self.dfs(image, visited, color, row-1, col, target)
        self.dfs(image, visited, color, row, col+1, target)
        self.dfs(image, visited, color, row, col-1, target)